test/test_main.py

```diff
@@ -261,69 +262,63 @@ def test_event_not_implemented(self):
             self.fail('Event simulation for an unsupported event'
                       ' results in a non-zero exit code returned')
 
-    def test_setup_hooks(self):
+    def test_setup_event_links(self):
         """Test auto-creation of symlinks for supported events.
         """
-        event_hooks = [f'hooks/{e.replace("_", "-")}'
-                       for e in charm.Charm.on.events().keys()
-                       if e != 'install']
-
-        install_link_path = JUJU_CHARM_DIR / 'hooks/install'
-
-        # The symlink is expected to be present in the source tree.
-        self.assertTrue(install_link_path.exists())
-        # It has to point to main.py in the lib directory of the charm.
-        self.assertEqual(os.readlink(install_link_path), self.CHARM_PY_RELPATH)
-
-        def _assess_setup_hooks(event_spec):
-            event_hook = JUJU_CHARM_DIR / f'hooks/{event_spec.event_name}'
-
-            # Simulate a fork + exec of a hook from a unit agent.
-            self._simulate_event(event_spec)
-
-            r, _, files = next(os.walk(JUJU_CHARM_DIR / 'hooks'))
-
-            self.assertTrue(event_spec.event_name in files)
-
-            for event_hook in event_hooks:
-                self.assertTrue(os.path.exists(event_hook))
-                self.assertEqual(os.readlink(event_hook), 'install')
-                self.assertEqual(os.readlink('hooks/install'),
-                                 self.CHARM_PY_RELPATH)
+        event_hooks = [f'hooks/{e.replace("_", "-")}' for e in charm.Charm.on.events().keys()]
 
         charm_config = base64.b64encode(pickle.dumps({
             'STATE_FILE': self._state_file,
         }))
 
-        # Assess 'install' first because upgrade-charm or other
-        # events cannot be handled before install creates symlinks for them.
-        events_to_assess = (
-            EventSpec(InstallEvent, 'install', charm_config=charm_config),
-            EventSpec(StartEvent, 'start', charm_config=charm_config),
+        initial_combinations = {
+            'hooks_iaas': (
+                # 'install' comes first because upgrade-charm or other events cannot be handled before install creates symlinks for them.
+                EventSpec(InstallEvent, 'install', charm_config=charm_config),
+            ),
+            'hooks_iaas_storage': (
+                # Storage hooks run before "install" so this case needs to be checked as well.
+                EventSpec(InstallEvent, 'install', charm_config=charm_config),
+                EventSpec(StorageAttachedEvent, 'disks-storage-attached', charm_config=charm_config),
+            ),  # Similar to the above test cases for Kubernetes charms which do not get the install event executed (see LP: #1854635).
+            'hooks_caas': (
+                EventSpec(StartEvent, 'start', charm_config=charm_config),
```

Comment:
Same.. per the conversation yesterday, this test is still confusing the idea of "this might happen" with "this is being tested". We need to make more clear what is being tested here, and I don't think we need to verify whether all potential combinations would work.

For example, some thing I can actually come up while brainstorming on this:

1. Is an event that is supposed to setup links actually doing that?
2. Is such an event idempotent if run multiple times?

What else?

If indeed the checks look that simple and objective, then this whole test ought to be simpler, more readable, and more objective than it currently looks.

Reply:
> I don't think we need to verify whether all potential combinations would work

What I had in mind:

All **valid combinations** of initial symlinks that a developer might include into the charm must result in a correct outcome: **all** event symlinks are present after **any** event execution with initial events executed first and other events executed afterwards.

We do not unlink anything in the hooks directory anymore so checking that all symlinks are present after events that are not initial (install, start, *-storage-attached) is not as important anymore. However, I would still do it for some events to avoid having a test that knows the internals and to make sure other events do not break the invariant.

Reply:
Without other events: https://github.com/canonical/operator/pull/63/commits/cec63e32c51fe54251a9f42035a49babe41628e1